loadfile/loadfile.go

```diff
@@ -7,41 +7,168 @@ import (
+			AbsolutePath: abspath,
+		}
+	}
+	return &fileCache{
+		rootDir: absDir, files: filesAbsPaths}, nil
+}
+
+// NewFileCache creates a file cache where the fileContents is
+// a mapping of paths to file contents. The keys in fileContents
+// become the keys in Files, and the absolute paths and ID for
+// each context file.
+func NewFileCache(rootDir string, fileContents map[string][]byte) FileCache {
+	files := map[string]ContextFile{}
+	for key, content := range fileContents {
+		files[key] = ContextFile{
+			ID:           key,
+			AbsolutePath: key,
+			Content:      content,
+		}
+	}
+	return &fileCache{
+		rootDir: rootDir,
+		files:   files,
+	}
+}
+
+// LoadContext reads the content of each context file into Content.
+func (fc *fileCache) LoadContext() error {
+	result := map[string]ContextFile{}
+	var err error
+	for key, cf := range fc.files {
+		absPath := cf.AbsolutePath
+		fileData, err := os.ReadFile(filepath.Clean(absPath))
+		if err != nil {
+			return fmt.Errorf("error reading file %s (%w)", absPath, err)
+		}
+		result[key] = ContextFile{
+			ID:           cf.ID,
+			AbsolutePath: cf.AbsolutePath,
+			Content:      fileData,
+		}
+	}
+	fc.files = result
+	return err
+}
+
+// RootDir returns the root directory used by files with
+// relative file paths.
+func (fc *fileCache) RootDir() string {
+	return fc.rootDir
+}
+
+// GetByKey asks if the file cache contains the given fileKey.
+func (fc *fileCache) GetByKey(fileKey string) (*ContextFile, bool) {
+	cf, ok := fc.files[fileKey]
+	return &cf, ok
+}
+
+// AbsPathByKey returns the absolute file path of a given file key,
+// if it exists in the file cache, nil otherwise.
+func (fc *fileCache) AbsPathByKey(fileKey string) *string {
+	cf, ok := fc.GetByKey(fileKey)
+	if !ok {
+		return nil
+	}
+	return &cf.AbsolutePath
+}
+
+// ContentByKey returns the file content of a given file key, if it
+// exists in the file cache, nil otherwise.
+func (fc *fileCache) ContentByKey(fileKey string) []byte {
+	cf, ok := fc.GetByKey(fileKey)
+	if !ok {
+		return nil
+	}
+	return cf.Content
+}
+
+// Contents returns a mapping of the file cache's file keys to file Content.
+func (fc *fileCache) Contents() map[string][]byte {
+	result := map[string][]byte{}
+	for key, f := range fc.files {
+		result[key] = f.Content
+	}
+	return result
+}
+
+// AbsPaths returns a mapping of the file cache's file keys to
+// absolute file paths.
+func (fc *fileCache) AbsPaths() map[string]string {
+	result := map[string]string{}
+	for key, f := range fc.files {
+		result[key] = f.AbsolutePath
+	}
+	return result
+}
+
+// Files returns the mapping of file keys to their ContextFile.
+func (fc *fileCache) Files() map[string]ContextFile {
+	return fc.files
+}
+
+// MergeFileCaches merges any number of file caches into one file cache.
+// The new root directory will be the root directory of the last file cache
+// argument. File keys found later in iteration will overwrite previously
+// file keys, if there is a name clash.
```

Comment:
Both aspects of this seem potentially problematic, given the "required files" keys you start with, since we'd expect them to be overwritten. If your merge isn't going to deal with that until you add sub-workflows, maybe merge itself should be deferred. Simply overwriting the root also seems potentially problematic, depending on how and where its used, since it'll no longer necessarily even be a component of some of the absolute paths.

Something like a chain of `FileCache` objects might make more sense to nest workflow contexts: a subworkflow would then search its own and fall back to the parent(s) (hierarchically) if necessary, instead of overwriting the parent context. When when you finish a workflow, you can just pop its context and throw it away without having irrevocably modified the parent context.

Reply:
Merging the `FileCache`s seems like the path to least change for the rest of this repo's code base, and this merge function has similar overwrite precedence as other hash map merge implementations that I've seen. The `RootDir` seemed like it should follow the same precedence. This is an instance where `FileCache` would be coupled to the Arcaflow Engine because I'm relying on there only ever being given one root directory. I think it's an acceptable coupling for this PR.

I would also need help understanding how to implement this hierarchical stack of file caches that you're talking about.

Reply:
Given that it doesn't seem to be referenced in this PR (outside of test code), it seems like it should be deferred to another change (if, indeed, it is actually needed there).  (I won't try to explicate the hierarchical stack -- there is no room in the margin here for that proof. 😉)